File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/EnteredTextViewModel.java
package betrip.kitttn.architecturecomponentstest.vm;

import android.arch.lifecycle.ViewModel;

import io.reactivex.Observable;
import io.reactivex.subjects.PublishSubject;

/**
 * @author kitttn
 */

public class EnteredTextViewModel extends ViewModel {
    private PublishSubject<String> enteredText = PublishSubject.create();
    private String lastQuery = "";

    public Observable<String> getEnteredText() {
        return enteredText;
    }

    public String getLastQuery() {
        return lastQuery;
    }

    public void textChanged(String newText) {
        this.lastQuery = newText;
        this.enteredText.onNext(newText);
    }
}

File: app/src/main/java/betrip/kitttn/architecturecomponentstest/vm/EnteredTextViewModel.kt
package betrip.kitttn.architecturecomponentstest.vm

import android.arch.lifecycle.ViewModel
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.subjects.PublishSubject

/**
 * @author kitttn
 */

class EnteredTextViewModel : ViewModel() {
    private val enteredText = PublishSubject.create<String>()
    private var lastQuery = ""

    fun getEnteredText(): Observable<String> = enteredText

    fun getLastQuery(): Single<String> {
        return Single.just(lastQuery)
    }

    fun textChanged(enteredText: String) {
        lastQuery = enteredText
        this.enteredText.onNext(enteredText)
    }
}